ambry-cloud/src/main/java/com/github/ambry/cloud/azure/CosmosDataAccessor.java

```diff
@@ -228,38 +230,46 @@ void testConnectivity() {
   }
 
   /**
-   * Upsert the blob metadata document in the CosmosDB collection.
+   * Upsert the blob metadata document in the CosmosDB container asynchronously.
    * @param blobMetadata the blob metadata document.
-   * @return the {@link CosmosItemResponse} containing {@link CloudBlobMetadata} returned by the operation, if successful.
-   * @throws CosmosException if the operation failed.
+   * @return a {@link CompletableFuture} that will eventually contain either the upserted {@link CloudBlobMetadata} if
+   *         successful or {@link CosmosException} exception if an error occurred.
    */
-  CosmosItemResponse<CloudBlobMetadata> upsertMetadata(CloudBlobMetadata blobMetadata) throws CosmosException {
-    return executeCosmosAction(
-        () -> cosmosAsyncContainer.upsertItem(blobMetadata, new PartitionKey(blobMetadata.getPartitionId()),
-            new CosmosItemRequestOptions()).block(), azureMetrics.documentCreateTime);
+  CompletableFuture<CloudBlobMetadata> upsertMetadataAsync(CloudBlobMetadata blobMetadata) {
+
+    Timer.Context operationTime = azureMetrics.documentCreateTime.time();
+    return cosmosAsyncContainer.upsertItem(blobMetadata, new PartitionKey(blobMetadata.getPartitionId()),
+        new CosmosItemRequestOptions()).toFuture().handle((response, throwable) -> {
```

Comment:
what is the purpose of using handle, if we just going to rethrow the exception
```
toFuture().thenApplyAsync(CosmosItemResponse::getItem);
```
would just do the work.

Reply:
I was wondering that if current completion stage completes exceptionally, then code to stop the operation timer inside `thenApply` completion stage wouldn't be executed. Hence, used `handle` method.